ocaml/xapi/storage_impl.ml

```diff
@@ -189,11 +190,42 @@ module Vdi = struct
     else
       Vdi_automaton.Detached
 
-  let set_dp_state dp state t =
-    let rest = List.filter (fun (u, _) -> u <> dp) t.dps in
+  (** [get_dp_vm] returns [vm] of type Vm.t. If the mapping between a datapath and vm exists, it will return that vm. otherwise it will attempt to claculate it. If that is not pisslbe it will return a blank vm *)
+
+  let get_dp_vm dp t =
+    if dp_on_vdi dp t then
+      let _vm = List.assoc dp t.dpv in
+      match s_of_vm _vm with
+      | "" ->
+          (* Blank vms will occur if host is being upgraded. these should not normally occur*)
+          debug
+            "[Vdi.get_dp_vm] Caught blank vm on dp:%s. To remove blank \
+             references: restart host"
+            dp ;
+          vm_of_s ""
+      | vm ->
+          vm_of_s vm
+    (* This else block should only be called if there is no dp in the dpv list. This should only be an issue during the upgrade*)
+    else
+      let vm =
+        debug
+          "[Vdi.get_dp_vm] No vm was found on the datapath: trying to obtain \
+           vm from parsing dp %s"
+          dp ;
+        try Scanf.sscanf dp "vbd/%8i/%_s" string_of_int
```

Comment:
I wouldn't bother with this. Many of the dp strings are not formatted this way, so it will only work some of the time, which doesn't buy us much.